concurrency/smallTest/SmallAtomViorOrOverSync.java

```diff
@@ -0,0 +1,190 @@
+        return myMap.get("lala");
+    }
+    
+    // ========= check and put =========
+    public void atomViol01() {
+        String myVal = myMap.get(key);
+        if (myVal == null) {
+            myMap.put(key, val);
+        }
+    }
+    
+    public void atomViol02() {
+        if (myMap.get(key) == null) {
+            myMap.put(key, val);
+        }
+    }
+    
+    public void atomViol03() {
+        if ( ! myMap.containsKey(key)) {
+            myMap.put(key, val);
+        }
+    }
+    
+    // =========== check and get ===========
+    
+    public void atomViol04() {
+        String myVal = myMap.get(key);
+        if (myVal != null) {
+            String myVal = myMap.get(key);
+            System.out.println(myVal.length()); // ensure accesses null
+        }
+    }
+    
+    public void atomViol05() {
+        if (myMap.get(key) != null) {
+            String myVal = myMap.get(key);
+            System.out.println(myVal.length()); // ensure accesses null
+        }
+    }
+    
+    public void atomViol06() {
+        if ( myMap.containsKey(key)) {
+            String myVal = myMap.get(key);
+            System.out.println(myVal.length()); // ensure accesses null
+        }
+    }
+    
+    // ================ check and remove ===============
+    
+    public void atomViol_checkRemove_01() {
+        if (myMap.get(key) != null) {
+            String myVal = myMap.remove(key);
+            System.out.println(myVal.size());            
+        }
+    }
+    
+    // ========= check and put with additional info about the put =========
+    public void atomViol07() {
+        String myVal = myMap.get(key);
+        if (myVal == null) {
+            myMap.put(key, val);
+        }
+        return val;
+    }
+    
+    public void atomViol08() {
+        if (myMap.get(key) == null) {
+            myMap.put(key, val);
+        }
+        System.out.println(val);
+    }
+    
+    public void atomViol09() {
+        if ( ! myMap.containsKey(key)) {
+            myMap.put(key, val);
+        }
+        System.out.println(val);
+        return val;
+    }
+    
+    // ================== over sync with put ================
+ 
+    public void atomViol07() {
+        String myKey = "lala" + 88;
+        String sillyVal = "lala" + 99;
+        synchronized(this) {
+            String localVal = myMap.get(myKey);
+            if (localVal == null) {
+                myMap.put(myKey, sillyVal);
+            }
+            return myKey;
+        }
+    }
+    
+    public synchronized void atomViol08() {
+        String myKey = "lala" + 88;
+        String sillyVal = "lala" + 99;
+        if (myMap.get(myKey) == null) {
+            myMap.put(myKey, sillyVal);
+        }
+    }
+    
+    public String atomViol09() {
+        String myKey = "lala" + 88;
+        String sillyVal = "lala" + 99;
+        synchronized (this) {
+            if ( ! myMap.containsKey(myKey)) {
+                myMap.put(myKey, sillyVal);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 These lines use `ConcurrentHashMap`&#39;s **sequential** API and are protected by `synchronized` to ensure they execute together: **122 and 123**. This synchronization may reduce performance by limiting parallelism and does not take advantage of `ConcurrentHashMap`&#39;s **concurrent** API.

**Fix**  
 Consider replacing the check and `put()` with `putIfAbsent()` and then removing the synchronization.

**More info**  
 `putIfAbsent()` is thread-safe and puts the value only if the `ConcurrentHashMap` does not contain the key. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/HistoryFileManager.java#L927-L932) (external link).